src/nnet3/nnet-am-decodable-simple.h

```diff
@@ -392,6 +392,170 @@ class DecodableAmNnetSimpleParallel: public DecodableInterface {
+  // When this function is triggered. We always call Compute() and 
+  // GetFinishedUtterance to reduce the load of the class.
+  inline bool Overload() const { return utt_list_.size() > max_capacity_; }
+
+  // It takes features as input, and you can either supply a
+  // single iVector input, estimated in batch-mode ('ivector'), or 'online'
+  // iVectors ('online_ivectors' and 'online_ivector_period', or none at all.
+  void AcceptInput(const std::string &utt_id,
+                   const MatrixBase<BaseFloat> &feats,
+                   const VectorBase<BaseFloat> *ivector = NULL,
+                   const MatrixBase<BaseFloat> *online_ivectors = NULL);
+  
+  
+  // Gets the output for a finished utterance. This is a non-blocking call.
+  bool GetFinishedUtterance(std::string *uttid, Matrix<BaseFloat> *output_matrix);
+
+  // This is a blocking call. When we call this function, we try to finish
+  // computing all the utterances which are still stored in this class.
+  // If 'flush == true', it would ensure that even if a batch wasn't ready,
+  // the computation would be run. When all the feature matrixes are read,
+  // we always call it one time to deal with the remain.
+  // This function is also useful in overload case.
+  void Compute(bool flush);
+
+ private:
+  KALDI_DISALLOW_COPY_AND_ASSIGN(NnetSimpleBatch);
+
+  // When an utterance is taken by GetFinishedUtterance(), clear its information.
+  void Clear(std::string utt_id);
+
+  // According to the information in batch_info_queue_, prepare the 'batch' data,
+  // ComputationRequest , compute and get the results out.
+  void DoNnetComputation();
+
+  // Gets the iVector that will be used for this chunk of frames, if we are
+  // using iVectors (else does nothing).  note: the num_output_frames is
+  // interpreted as the number of t value, which in the subsampled case is not
+  // the same as the number of subsampled frames (it would be larger by
+  // opts_.frame_subsampling_factor).
+  void GetCurrentIvector(std::string utt_id,
+                         int32 output_t_start,
+                         int32 num_output_frames,
+                         Vector<BaseFloat> *ivector);
+
+  // called from constructor
+  void CheckAndFixConfigs();
+
+  // called from AcceptInput()
+  void CheckInput(const std::string &utt_id,
+                  const MatrixBase<BaseFloat> &feats,
+                  const VectorBase<BaseFloat> *ivector = NULL,
+                  const MatrixBase<BaseFloat> *online_ivectors = NULL);
+
+  // called from AcceptInput() or Compute(). Prepare the batch_info_queue_ which
+  // will be used to compute.
+  void PrepareBatchInfo();
+
+
+  NnetSimpleComputationOptions opts_;
+  const Nnet &nnet_;
+  int32 nnet_left_context_;
+  int32 nnet_right_context_;
+  int32 output_dim_;
+  // the log priors (or the empty vector if the priors are not set in the model)
+  CuVector<BaseFloat> log_priors_;
+
+  std::unordered_map<std::string, Matrix<BaseFloat>> feats_;
```

Comment:
You should never store things like Matrix or Vector in an unordered_map directly, you should store pointers to them.
The issue is that any time the size of the map changes, they will have to be copied which is a fairly big operation.  Obviously you'll have to delete the pointers any time you delete a key from the map.